CenterForOpenScience/osf.io
# [Hotfix] Add BaseFileNode index for target/provider/type

## Purpose
Avoid costly BitmapHeap Scan when doing things like `OsfStorageFolder.objects.filter(target=node)`, etc.

## Changes
* Alter `index_together`

## QA Notes
See if files load any faster? No direct UI changes, may not affect the generic file page queries (especially not until the table is vacuumed, which happens automatically over long periods of time). Added specifically to further optimize https://github.com/CenterForOpenScience/osf.io/pull/8849

## Side Effects
None expected, other than what's listed below

## Deployment Notes
Because this is decently-sized index on one of our largest tables, this **MUST NOT** be deployed without downtime. Indices require exclusive table locks while being created*, therefore this requires maintenance downtime. 

\* Except concurrently, which is [technically possible](https://docs.djangoproject.com/en/2.1/ref/migration-operations/#separatedatabaseandstate) but not [supported by Django](https://code.djangoproject.com/ticket/21039). Doing so would require the migration operation to be something like:
```
migrations.SeparateDatabaseAndState(
    state_operations=[
        migrations.AlterIndexTogether(
            model_name='basefilenode',
            index_together=set([('target_content_type', 'target_object_id'), ('target_content_type', 'target_object_id', 'provider', 'type')]),
        ),
    ],
    database_operations=[
        migrations.RunSQL([
            'CREATE INDEX CONCURRENTLY "osf_basefilenode_ttid_tcid_prv_type_deadbeef_idx" ON "osf_basefilenode" ("target_object_id", "target_content_type_id", "provider", "type");',
        ], [
            'DROP INDEX IF EXISTS "osf_basefilenode_ttid_tcid_prv_type_deadbeef_idx", RESTRICT;'
        ])
    ]
)
```
However, due to [caveats](https://www.postgresql.org/docs/current/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY), it's possible that index would not be successfully created, leading to increased overhead and none of the benefits, and possible deadlock during the failed creation process.

## Ticket
None known
